### PR TITLE
Sarkar/growing bucket for beam

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
Extends growing bucket optimization from  greedy to beam.

Original PR for growing bucket optimization for greedy [here](https://github.com/huggingface/optimum-habana/commit/6f5f97f0e19755ab6ba29d496fdd9c58a785ac34)

Testing in progress

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -296,6 +319,55 @@ def _get_stopping_criteria(
         criteria = self._merge_criteria_processor_list(criteria, stopping_criteria)
         return criteria
 
+    @torch.no_grad()
```
Comment (ssarkar2) [inline optimum/habana/transformers/generation/utils.py:322]: Line 283:

if (
                generation_config.static_shapes
                and self.config.is_encoder_decoder
                and self.generation_config.generation_mode == GenerationMode.GREEDY_SEARCH
            ):
                criteria.append(StaticMaxLengthCriteria(generation_config.max_length))


Need to add condition for Beam as well

**Comment by ssarkar2** [comment]
@regisss : can you please take a look at this PR. This ports this [PR ](https://github.com/huggingface/optimum-habana/pull/417) from greedy to beam.
Perf/accuracy tests have been done internally, and they look good.

This is not necessary for 1.12, which I understand is your priority right now

File: examples/text-generation/README.md [hunk 2]
```diff
@@ -170,7 +170,7 @@ python run_generation.py \
 ### Using growing bucket optimization
 With `--bucket_size`, instead of padding up the kv-cache up to full size before starting, we grow the cache/input in multiples of `bucket_size`. This helps increase throughput and also reduce number of compilations if the dataset has varying prompt lengths.
 
-> For now, it is available only for greedy generation, and cannot be used with `--reuse_cache`.
+> For now, it is available only for greedy and beam generation, and cannot be used with `--reuse_cache`.
```
Comment (regisss) [inline examples/text-generation/README.md:173]:
```suggestion
> For now, it is available only for greedy and beam search generation, and cannot be used with `--reuse_cache`.
```
Let's add this to make sure users don't mix it up with beam sample.

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -488,9 +563,9 @@ def generate(
                 inputs_tensor, generation_config.pad_token_id, generation_config.eos_token_id
             )
 
-        is_greedy_and_bucket = (
-            generation_config.bucket_size > 0
-            and self._get_generation_mode(generation_config, assistant_model) == GenerationMode.GREEDY_SEARCH
+        is_greedyorbeam_and_bucket = generation_config.bucket_size > 0 and (
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:566]:
```suggestion
        is_greedy_or_beam_and_bucket = generation_config.bucket_size > 0 and (
```

File: optimum/habana/transformers/generation/utils.py [hunk 4]
```diff
@@ -502,7 +577,7 @@ def generate(
 
             if not self.config.is_encoder_decoder:
                 # only pad if bucket_size < -1. If we are bucketing (bucket_size > 0), then that is taken care in greedy_search()
-                if not is_greedy_and_bucket:
+                if not is_greedyorbeam_and_bucket:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:580]:
```suggestion
                if not is_greedy_or_beam_and_bucket:
```

File: optimum/habana/transformers/generation/utils.py [hunk 5]
```diff
@@ -593,7 +668,7 @@ def generate(
                 calculated_max_length = input_ids.shape[-1] + generation_config.max_new_tokens
             if generation_config.use_cache and generation_config.reuse_cache:
                 bs, _ = input_ids.shape
-                if not is_greedy_and_bucket:
+                if not is_greedyorbeam_and_bucket:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:671]:
```suggestion
                if not is_greedy_or_beam_and_bucket:
```

**Review by regisss — COMMENTED** [review]
LGTM! I just left a few minor comments.
Could you also share a command line that I can quickly run to test it please? @ssarkar2

**Comment by ssarkar2** [comment]
Commands to test:

**1.** Run with a dataset (--dataset_name squad --column_name context ) for 10 steps with bs=1 (--dataset_max_samples 10 --batch 1 ) without cropping prompt (--max_input_tokens -1)
beams is enabled by --num_beams 2 

 python run_generation.py --model_name_or_path path_to_model --use_hpu_graphs --use_kv_cache  --max_new_tokens 256 --batch 1  --num_beams 2 --dataset_name squad --column_name context --max_input_tokens -1 --dataset_max_samples 10

Adding "--bucket 50" to enable current PR code
 python run_generation.py --model_name_or_path path_to_model --use_hpu_graphs --use_kv_cache  --max_new_tokens 256 --batch 1  --num_beams 2 --dataset_name squad --column_name context --max_input_tokens -1 --dataset_max_samples 10 **--bucket 50**

the bucketed one is faster than the base one

**2.** Single prompt test:
python run_generation.py --model_name_or_path path_to_model --use_hpu_graphs --use_kv_cache  --max_new_tokens 256 --batch 16 --num_beams 2

python run_generation.py --model_name_or_path path_to_model  --use_hpu_graphs --use_kv_cache  --max_new_tokens 256 --batch 16 --num_beams 2 **--bucket 50**

Here we may see some improvement, but maybe not too much difference. Could depend on model, promptlen. max_new_tokens etc

This change helps reduce num compilations, so it is more useful in test 1 (dataset, with different input shapes)

models tested: opt-350m, llama-7b, bloom-7b

**Comment by regisss** [comment]
> @regisss , i see some distributed tests fail in "Unit and integration tests / Run tests for optimum.habana.transformers (pull_request_target)". Are these expected failures, or introduced by my PR?

I've seen these errors in the CI of another PR that was not related to distributed training either. That's a bit weird but that's independent of your PR. I'm going to relaunch these tests.

File: examples/text-generation/run_generation.py [hunk 6]
```diff
@@ -569,7 +569,7 @@ def preprocess_function(examples):
         if prompt_length <= 0:
             # Todo please check if this collate function is suitable for your model
             # This has been tested for OPT, and Bloom
```
Comment (regisss) [inline examples/text-generation/run_generation.py:571]: We should update this comment too and add Llama I guess

**Review by regisss — COMMENTED** [review]
I just spotted one comment to update. I'm going to test it quickly and then it should be good to merge!

**Comment by ssarkar2** [comment]
@regisss can we merge this?

**Review by regisss — APPROVED** [review]
LGTM!
